frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -587,18 +605,87 @@ impl<T: Config> Pallet<T> {
 						Ok(now)
 					}
 				}
-			});
+			},
+		);
 
 		match mutate_stat {
 			// all good
 			Ok(Ok(_)) => Ok(()),
 			// failed to write.
-			Ok(Err(_)) => Err(MinerError::Lock("failed to write to offchain db.")),
+			Ok(Err(_)) => Err(MinerError::Lock("failed to write to offchain db [repeat].")),
 			// fork etc.
 			Err(why) => Err(MinerError::Lock(why)),
 		}
 	}
 
+	/// Ensure a new offchain worker is not being executed at this point in time, by checking
+	/// `OFFCHAIN_RUNNING` storage.
+	///
+	/// If `OK(_)` is returned, `true` is written to `OFFCHAIN_RUNNING`, which in turn prevents any
+	/// further call to this function to succeed.
+	///
+	/// It is up to the call site, and ABSOLUTELY CRUCIAL to make sure `release_offchain_lock` is
+	/// called appropriately to ensure starvation does not occur.
+	pub(crate) fn ensure_offchain_not_running() -> Result<(), MinerError> {
+		let guard = StorageValueRef::persistent(&OFFCHAIN_RUNNING);
+
+		let mutate_stat =
+			guard.mutate::<_, &'static str, _>(|maybe_running: Option<Option<bool>>| {
+				match maybe_running {
+					Some(Some(false)) => {
+						// no one else is running. Set it, and return Ok.
+						Ok(true)
+					}
+					Some(Some(true)) => {
+						// another thread is currently running.
+						Err("still running.")
```

Comment:
This may cause issues if node quites unexpectedly in the middle of execution - the off-chain DB may be left in "running" state and never really release the lock.

Have you considered using `StorageLock` structure? You could set the auto-unlock sufficiently high and also use the guard to "prolong" the lock in case the worker is still running.

Reply:
ahhh, no, I didn't know about `StorageLock`, and looking at its API, I should use that instead. With a long enough auto-unlock (i.e. one session length) it should behave pretty well.